### DISCUSSION TITLE
as_xyz accessors should return Result

### DISCUSSION THREAD

**Discussion by rcoh** [smithy-lang/smithy-rs#387] [Ideas]
Currently, `as_xyz` accessors return `Option<T>`. But this can be confusing for users when the type is actually a different type. Instead, `as_xyz` should return `Result<&T, &Self>`:

Proposal:
```rust
    pub fn as_ss(&self) -> Result<&std::vec::Vec<std::string::String>, &Self> {
        if let AttributeValue::Ss(val) = &self {
            Ok(&val)
        } else {
            Err(&self)
        }
    }
```

This should make the problem immediately clear and easy to debug.

**Comment by robisonsantos** [comment]
On the same methods, it would be nice if the method "respected" the type from dynamo and returned the proper representation. For example, `as_n()` is always handling a number, `as_b()` is always handling  a boolean, etc.

If we take the "principle of least surprise", I'd expect `as_n` to return a `Result<float, &Self>`, `as_b` to return a `Result<bool, &Self>` etc.

Consumers of the API should not need to parse and transform data types read from dynamo if those types are well known.

**Reply by rcoh** [reply]
This is a little trickier but may still be feasible—what should the rule be? I guess we could auto-unwrap through structures with one field?

Normally backwards compatibility problems would prevent this but unions are actually much more stringent

**Comment by rcoh** [comment]
This is fixed in #527
